File: api/cron-ingest.py
import os
import urllib.request
import json
from flask import Flask, request, jsonify

app = Flask(__name__)

@app.route('/api/cron-ingest', methods=['GET', 'POST'])
@app.route('/', methods=['GET', 'POST'])
def cron_ingest():
    auth_header = request.headers.get('Authorization', '')
    cron_secret = os.environ.get('CRON_SECRET')

    # Validate Authorization header if CRON_SECRET is configured
    if cron_secret and auth_header != f"Bearer {cron_secret}":
        return jsonify({"error": "Unauthorized. Invalid Cron Secret."}), 401

    deploy_hook_url = os.environ.get('VERCEL_DEPLOY_HOOK_URL')
    if not deploy_hook_url:
        print("[CRON] VERCEL_DEPLOY_HOOK_URL environment variable is missing.")
        return jsonify({"error": "VERCEL_DEPLOY_HOOK_URL environment variable is not configured."}), 500

    try:
        print("[CRON] Triggering rebuild via Vercel Deploy Hook...")
        req = urllib.request.Request(deploy_hook_url, method='POST')
        with urllib.request.urlopen(req) as response:
            res_data = response.read().decode('utf-8')
            data = json.loads(res_data)
            print(f"[CRON] Deploy Hook triggered successfully. Job ID: {data.get('job', {}).get('id')}")
            return jsonify({
                "success": True,
                "message": "Daily rebuild and ingest pipeline triggered successfully.",
                "jobId": data.get('job', {}).get('id')
            }), 200
    except Exception as e:
        print(f"[CRON] Error triggering Deploy Hook: {e}")
        return jsonify({
            "error": "Failed to trigger Vercel Deploy Hook rebuild.",
            "details": str(e)
        }), 500

if __name__ == '__main__':
    app.run(port=3000)

File: api/cron-ingest.js
/**
 * Vercel Cron handler to trigger daily RAG index updates at 10:00 AM IST (04:30 AM UTC).
 * It sends a request to the Vercel Deploy Hook, triggering a rebuild.
 * During the rebuild, the build script crawls the AMC pages and regenerates the vector index.
 */

export default async function handler(req, res) {
  // Validate request is triggered by Vercel Cron
  // Vercel Cron sends a Bearer token in the Authorization header
  const authHeader = req.headers.authorization;
  const cronSecret = process.env.CRON_SECRET;

  // Only validate if CRON_SECRET is configured in the environment
  if (cronSecret && authHeader !== `Bearer ${cronSecret}`) {
    return res.status(401).json({ error: 'Unauthorized. Invalid Cron Secret.' });
  }

  const deployHookUrl = process.env.VERCEL_DEPLOY_HOOK_URL;
  if (!deployHookUrl) {
    console.warn('[CRON] VERCEL_DEPLOY_HOOK_URL environment variable is missing.');
    return res.status(500).json({ 
      error: 'VERCEL_DEPLOY_HOOK_URL environment variable is not configured.' 
    });
  }

  try {
    console.log('[CRON] Triggering rebuild via Vercel Deploy Hook...');
    
    const response = await fetch(deployHookUrl, {
      method: 'POST'
    });

    if (!response.ok) {
      throw new Error(`Deploy Hook returned status: ${response.status}`);
    }

    const data = await response.json();
    console.log('[CRON] Deploy Hook triggered successfully. Job ID:', data.job?.id);

    return res.status(200).json({
      success: true,
      message: 'Daily rebuild and ingest pipeline triggered successfully.',
      jobId: data.job?.id
    });

  } catch (error) {
    console.error('[CRON] Error triggering Deploy Hook:', error);
    return res.status(500).json({ 
      error: 'Failed to trigger Vercel Deploy Hook rebuild.',
      details: error.message 
    });
  }
}
